pkg/controller/sync/aggregator/aggregator.go

```diff
@@ -0,0 +1,105 @@
+package aggregator
+
+import (
+	"fmt"
+	gosync "sync"
+
+	"golang.org/x/exp/maps"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type KindName struct {
+	Kind string
+	Name string
+}
+
+func NewGVKAggregator(allowedKeyKinds []string) *GVKAgreggator {
+	return &GVKAgreggator{
+		store:        make(map[KindName]map[schema.GroupVersionKind]struct{}),
+		reverseStore: make(map[schema.GroupVersionKind]map[KindName]struct{}),
+
+		allowedKeyKinds: allowedKeyKinds,
+	}
+}
+
+// todo comments.
+type GVKAgreggator struct {
+	mu gosync.RWMutex
+
+	store        map[KindName]map[schema.GroupVersionKind]struct{}
+	reverseStore map[schema.GroupVersionKind]map[KindName]struct{}
+
+	allowedKeyKinds []string
+}
+
+func (b *GVKAgreggator) validateKind(k KindName) error {
+	if len(b.allowedKeyKinds) == 0 {
+		return nil
+	}
+
+	for _, kk := range b.allowedKeyKinds {
+		if k.Kind == kk {
+			// we found a match so allow the kind
+			return nil
+		}
+	}
+
+	return fmt.Errorf("unsupported key kind for key (%s, %s); supported values are: %s", k.Kind, k.Name, b.allowedKeyKinds)
+}
+
+func (b *GVKAgreggator) ListGVKs() []schema.GroupVersionKind {
```

Comment:
We may want to have this return a set, since this list is meant to be a set.  Let's see if any other reviewer care.

Reply:
Hard to say what it should return (or whether this should exist) without knowing what functionality this is intended to support.

If this is intended for supporting "cache manager thing", set logic will be necessary in order to easily calculate deltas (new GVKs to watch, old GVKs that should no longer be watched).

IMO I'd avoid creating anything intended to support the "cache manager thing" until it's time to write "cache manager thing".

Reply:
I'll defer to Max here.  This method is like 3 lines so easy to change our mind.

r.e. the set vs. slice, my hunch is set is better but again it's easy to change later.

Reply:
IIUC, I think Max is asking us to defer the decision here entirely until we actually need it in the "cache manager thing". To that end, I've removed the `ListGVKs` func for now.